sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala

```diff
@@ -130,6 +132,25 @@ private[sql] object PhysicalRDD {
       metadata: Map[String, String] = Map.empty): PhysicalRDD = {
     // All HadoopFsRelations output UnsafeRows
     val outputUnsafeRows = relation.isInstanceOf[HadoopFsRelation]
-    PhysicalRDD(output, rdd, relation.toString, metadata, outputUnsafeRows)
+
+    val bucketSpec = relation match {
+      case r: HadoopFsRelation if relation.sqlContext.conf.bucketingEnabled() => r.bucketSpec
+      case _ => None
+    }
+
+    def toAttribute(colName: String): Attribute = output.find(_.name == colName).get
+
+    val bucketedPhysicalRDD = for {
```

Comment:
Can you make this less functional?

something like:
val bucketedPhysicalRDD = bucketSpec.map { spec =>
  numBuckets = spec.numBuckets
  bucketColumns = spec.bucketColumnNames.map(toAttribute)
  val partitioning = HashPartitioning(bucketColumns, numBuckets)
  PhysicalRDD(output, rdd, relation.toString, metadata, outputUnsafeRows, partitioning)
}.getOrElse {
  PhysicalRDD(output, rdd, relation.toString, metadata, outputUnsafeRows)
}